core/src/main/java/com/linecorp/armeria/client/retry/RetryingRpcClient.java

```diff
@@ -124,9 +141,13 @@ private void doExecute0(ClientRequestContext ctx, RpcRequest req,
         final RpcResponse res = executeWithFallback(unwrap(), derivedCtx,
                                                     (context, cause) -> RpcResponse.ofFailure(cause));
 
+        final RetryConfig<RpcResponse> retryConfig = mapping().get(ctx, req);
+        final RetryRuleWithContent<RpcResponse> retryRule =
+                retryConfig.needsContentInRule() ?
+                retryConfig.retryRuleWithContent() : RetryRuleUtil.fromRetryRule(retryConfig.retryRule());
```

Comment:
`RetryingRpcClient` only use `retryRuleWithContent` because there is no performance degression  using `RetryRuleWithContent`.

Could add a validation code to `builder(RetryConfigMapping<RpcResponse> mapping)` whether `retryRuleWithContent` is configured?

Reply:
It's impossible to validate if mapping would map to a RetryRule in one of the cases. RetryConfigMapping lets the use implement a factory that returns a RetryConfig for a given context. The RetryConfig could either have a RetryRule or a RetryRuleWithContent, and there is no way to check ahead of time if the user-defined factory would never produce a config with a RetryRuleWithContent for any given context.

The current implementation allows the mapping to have a RetryRule even if it's used in RpcClient, and it effectively turns it into RetryRuleWithContent with no problem.